precompile/contracts/ibc/contract.go

```diff
@@ -0,0 +1,2239 @@
+		consensusStateLen  - consensusState
+	*/
+	if remainingGas, err = contract.DeductGas(suppliedGas, createClientGas); err != nil {
+		return nil, 0, err
+	}
+	if readOnly {
+		return nil, remainingGas, vmerrs.ErrWriteProtection
+	}
+	// no input provided for this function
+
+	// Allow list is enabled and AllowFeeRecipients is a state-changer function.
+	// This part of the code restricts the function to be called only by enabled/admin addresses in the allow list.
+	// You can modify/delete this code if you don't want this function to be restricted by the allow list.
+	stateDB := accessibleState.GetStateDB()
+	// Verify that the caller is in the allow list and therefore has the right to call this function.
+	callerStatus := allowlist.GetAllowListStatus(stateDB, ContractAddress, caller)
+	if !callerStatus.IsEnabled() {
+		return nil, remainingGas, fmt.Errorf("non-enabled cannot call createClient: %s", caller)
+	}
+
+	nextClientSeq := uint64(0)
+	if accessibleState.GetStateDB().Exist(common.BytesToAddress([]byte("nextClientSeq"))) {
+		b := accessibleState.GetStateDB().GetPrecompileState(common.BytesToAddress([]byte("nextClientSeq")))
+		nextClientSeq = binary.BigEndian.Uint64(b)
+	}
+
+	// ClientStateBytes
+	carriage := uint64(0)
+	clientStateLen := new(big.Int).SetBytes(getData(input, carriage, 8)).Uint64()
+	carriage = carriage + 8
+
+	clientStateByte := getData(input, carriage, clientStateLen)
+
+	carriage = carriage + clientStateLen
+
+	clientState := &ibctm.ClientState{}
+	err = clientState.Unmarshal(clientStateByte)
+	if err != nil {
+		return nil, createClientGas, fmt.Errorf("error unmarshalling client state file: %w", err)
+	}
+
+	clientID := fmt.Sprintf("%s-%d", clientState.ClientType(), nextClientSeq)
+	nextClientSeq++
+
+	b := make([]byte, 8)
+	binary.BigEndian.PutUint64(b, nextClientSeq)
+	accessibleState.GetStateDB().SetPrecompileState(common.BytesToAddress([]byte("nextClientSeq")), b)
+
+	// ConsensusStateBytes
+	consensusStateLen := new(big.Int).SetBytes(getData(input, carriage, 8)).Uint64()
+	carriage = carriage + 8
+	consensusStateByte := getData(input, carriage, consensusStateLen)
+
+	consensusState := &ibctm.ConsensusState{}
+	err = consensusState.Unmarshal(consensusStateByte)
+	if err != nil {
+		return nil, 0, fmt.Errorf("error unmarshalling client state file: %w", err)
+	}
+	// store ClientStateBytes
+	clientStatePath := fmt.Sprintf("clients/%s/clientState", clientID)
+	accessibleState.GetStateDB().SetPrecompileState(common.BytesToAddress([]byte(clientStatePath)), clientStateByte)
+
+	// store ConsensusStateBytes
+	consensusStateByte, err = consensusState.Marshal()
+	if err != nil {
+		return nil, 0, errors.New("consensusState marshaler error")
+	}
+
+	consensusStatePath := fmt.Sprintf("clients/%s/consensusStates/%s", clientID, clientState.GetLatestHeight())
+	accessibleState.GetStateDB().SetPrecompileState(common.BytesToAddress([]byte(consensusStatePath)), consensusStateByte)
+	return []byte(clientID), createClientGas, nil
```

Comment:
In cosmos, calling createClient always emits an event. Perhaps this should emit a similar event?

Reply:
we added task to our backlog
https://consideritdone.atlassian.net/browse/AILC-116